// @nerimity/nerimity.js/build/Client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Button = exports.Post = exports.Message = exports.ServerChannel = exports.Channel = exports.Channels = exports.ServerMember = exports.ServerMembers = exports.ServerRole = exports.ServerRoles = exports.Server = exports.Servers = exports.Users = exports.Client = exports.Events = void 0;
const eventemitter3_1 = __importDefault(require("eventemitter3"));
const socket_io_client_1 = require("socket.io-client");
const EventNames_1 = require("./EventNames");
const RawData_1 = require("./RawData");
const MessageService_1 = require("./services/MessageService");
const serviceEndpoints_1 = require("./services/serviceEndpoints");
const PostsService_1 = require("./services/PostsService");
const ApplicationService_1 = require("./services/ApplicationService");
const ServerService_1 = require("./services/ServerService");
const bitwise_1 = require("./bitwise");
exports.Events = EventNames_1.ClientEvents;
class Client extends eventemitter3_1.default {
    constructor(opts) {
        super();
        if (opts?.urlOverride) {
            (0, serviceEndpoints_1.updatePath)(opts.urlOverride);
        }
        this.socket = (0, socket_io_client_1.io)(serviceEndpoints_1.path, {
            transports: ["websocket"],
            autoConnect: false,
        });
        this.channels = new Channels(this);
        this.users = new Users(this);
        this.servers = new Servers(this);
        this.posts = new Posts(this);
        new EventHandlers(this);
    }
    updateCommands(token, commands) {
        return (0, ApplicationService_1.updateCommands)(token, commands);
    }
    login(token) {
        this.token = token;
        this.socket.connect();
    }
}
exports.Client = Client;
class EventHandlers {
    constructor(client) {
        this.client = client;
        this.socket = client.socket;
        client.socket.on(EventNames_1.SocketServerEvents.CONNECT, this.onConnect.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.USER_AUTHENTICATED, this.onAuthenticated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_MEMBER_JOINED, this.onServerMemberJoined.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_MEMBER_UPDATED, this.onServerMemberUpdated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_MEMBER_LEFT, this.onServerMemberLeft.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_JOINED, this.onServerJoined.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_CHANNEL_CREATED, this.onServerChannelCreated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_CHANNEL_UPDATED, this.onServerChannelUpdated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_CHANNEL_DELETED, this.onServerChannelDeleted.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_LEFT, this.onServerLeft.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.MESSAGE_CREATED, this.onMessageCreated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.MESSAGE_BUTTON_CLICKED, this.onMessageButtonClicked.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_ROLE_CREATED, this.onServerRoleCreated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_ROLE_DELETED, this.onServerRoleDeleted.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_ROLE_UPDATED, this.onServerRoleUpdated.bind(this));
        client.socket.on(EventNames_1.SocketServerEvents.SERVER_ROLE_ORDER_UPDATED, this.onServerRoleOrderUpdated.bind(this));
    }
    onConnect() {
        this.socket.emit(EventNames_1.SocketClientEvents.AUTHENTICATE, {
            token: this.client.token,
        });
    }
    onAuthenticated(payload) {
        this.client.user = new ClientUser(this.client, payload.user);
        for (let i = 0; i < payload.servers.length; i++) {
            const server = payload.servers[i];
            this.client.servers.setCache(server);
        }
        for (let i = 0; i < payload.channels.length; i++) {
            const rawChannel = payload.channels[i];
            this.client.channels.setCache(rawChannel);
        }
        for (let i = 0; i < payload.serverMembers.length; i++) {
            const member = payload.serverMembers[i];
            this.client.users.setCache(member.user);
            const server = this.client.servers.cache.get(member.serverId);
            server?.members.setCache(member);
        }
        for (let i = 0; i < payload.serverRoles.length; i++) {
            const role = payload.serverRoles[i];
            const server = this.client.servers.cache.get(role.serverId);
            server?.roles.setCache(role);
        }
        this.client.emit(EventNames_1.ClientEvents.Ready);
    }
    onServerMemberJoined(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        this.client.users.setCache(payload.member.user);
        const member = server?.members.setCache(payload.member);
        if (!member)
            return;
        this.client.emit(EventNames_1.ClientEvents.ServerMemberJoined, member);
    }
    onServerMemberUpdated(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const member = server?.members.cache.get(payload.userId);
        if (!member)
            return;
        updateClass(member, payload.updated);
        this.client.emit(EventNames_1.ClientEvents.ServerMemberUpdated, member);
    }
    onServerJoined(payload) {
        const server = this.client.servers.setCache(payload.server);
        for (let i = 0; i < payload.members.length; i++) {
            const member = payload.members[i];
            this.client.users.setCache(member.user);
            server?.members.setCache(member);
        }
        for (let i = 0; i < payload.roles.length; i++) {
            const role = payload.roles[i];
            server?.roles.setCache(role);
        }
        for (let i = 0; i < payload.channels.length; i++) {
            const channel = payload.channels[i];
            this.client.channels.setCache(channel);
        }
        this.client.emit(EventNames_1.ClientEvents.ServerJoined, server);
    }
    onServerChannelCreated(payload) {
        const channel = this.client.channels.setCache(payload.channel);
        this.client.emit(EventNames_1.ClientEvents.ServerChannelCreated, channel);
    }
    onServerChannelUpdated(payload) {
        const channel = this.client.channels.cache.get(payload.channelId);
        const updated = payload.updated;
        if (channel) {
            updateClass(channel, updated);
            this.client.emit(EventNames_1.ClientEvents.ServerChannelUpdated, channel);
        }
    }
    onServerChannelDeleted(payload) {
        const channel = this.client.channels.cache.has(payload.channelId);
        if (channel) {
            this.client.channels.cache.delete(payload.channelId);
            this.client.emit(EventNames_1.ClientEvents.ServerChannelDeleted, {
                channelId: payload.channelId,
                serverId: payload.serverId,
            });
        }
    }
    onServerLeft(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        if (!server)
            return;
        this.client.emit(EventNames_1.ClientEvents.ServerLeft, server);
        this.client.servers.cache.delete(payload.serverId);
        this.client.channels.cache.forEach((channel) => {
            if (channel instanceof ServerChannel &&
                channel.serverId === payload.serverId) {
                this.client.channels.cache.delete(channel.id);
            }
        });
        server.members.cache.clear();
    }
    onServerMemberLeft(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const member = server?.members.cache.get(payload.userId);
        if (!member)
            return;
        this.client.emit(EventNames_1.ClientEvents.ServerMemberLeft, member);
        server?.members.cache.delete(payload.userId);
    }
    onMessageCreated(payload) {
        const message = new Message(this.client, payload.message);
        this.client.emit(EventNames_1.ClientEvents.MessageCreate, message);
    }
    onMessageButtonClicked(payload) {
        const button = new Button(this.client, payload);
        this.client.emit(EventNames_1.ClientEvents.MessageButtonClick, button);
    }
    onServerRoleCreated(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const role = server?.roles.setCache(payload);
        if (!role)
            return;
        this.client.emit(EventNames_1.ClientEvents.ServerRoleCreated, role);
    }
    onServerRoleDeleted(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const role = server?.roles.cache.get(payload.roleId);
        if (!role)
            return;
        server?.roles.cache.delete(payload.roleId);
        this.client.emit(EventNames_1.ClientEvents.ServerRoleDeleted, role);
    }
    onServerRoleUpdated(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        const role = server?.roles.cache.get(payload.roleId);
        if (!role)
            return;
        updateClass(role, payload.updated);
        this.client.emit(EventNames_1.ClientEvents.ServerRoleUpdated, role);
    }
    onServerRoleOrderUpdated(payload) {
        const server = this.client.servers.cache.get(payload.serverId);
        for (let i = 0; i < payload.roleIds.length; i++) {
            const roleId = payload.roleIds[i];
            const role = server?.roles.cache.get(roleId);
            if (!role)
                continue;
            role.order = i + 1;
        }
        this.client.emit(EventNames_1.ClientEvents.ServerRoleOrderUpdated, server?.roles);
    }
}
class Users {
    constructor(client) {
        this.client = client;
        this.cache = new Collection();
    }
    setCache(rawUser) {
        const user = new User(this.client, rawUser);
        this.cache.set(rawUser.id, user);
        return user;
    }
}
exports.Users = Users;
class Servers {
    constructor(client) {
        this.client = client;
        this.cache = new Collection();
    }
    setCache(rawServer) {
        const server = new Server(this.client, rawServer);
        this.cache.set(server.id, server);
        return server;
    }
}
exports.Servers = Servers;
class Server {
    constructor(client, server) {
        this.client = client;
        this.id = server.id;
        this.name = server.name;
        this.avatar = server.avatar;
        this.members = new ServerMembers(this.client);
        this.roles = new ServerRoles(this.client);
        this.defaultRoleId = server.defaultRoleId;
        this.createdById = server.createdById;
    }
    async banMember(userId) {
        return (0, ServerService_1.banServerMember)(this.client, this.id, userId);
    }
}
exports.Server = Server;
class ServerRoles {
    constructor(client) {
        this.client = client;
        this.cache = new Collection();
    }
    setCache(rawServerRole) {
        const server = new ServerRole(this.client, rawServerRole);
        this.cache.set(server.id, server);
        return server;
    }
}
exports.ServerRoles = ServerRoles;
class ServerRole {
    constructor(client, role) {
        this.client = client;
        this.server = this.client.servers.cache.get(role.serverId);
        this.id = role.id;
        this.name = role.name;
        this.permissions = role.permissions;
        this.hexColor = role.hexColor;
        this.order = role.order;
        this.isDefaultRole = this.server.defaultRoleId === this.id;
    }
}
exports.ServerRole = ServerRole;
class ServerMembers {
    constructor(client) {
        this.client = client;
        this.cache = new Collection();
    }
    setCache(rawMember) {
        const member = new ServerMember(this.client, rawMember);
        this.cache.set(member.user.id, member);
        return member;
    }
}
exports.ServerMembers = ServerMembers;
class ServerMember {
    constructor(client, member) {
        this.client = client;
        this.id = member.user.id;
        this.roleIds = member.roleIds;
        this.nickname = member.nickname;
        this.user = this.client.users.cache.get(member.user.id);
        this.server = this.client.servers.cache.get(member.serverId);
    }
    toString() {
        return `[@:${this.id}]`;
    }
    async ban() {
        return (0, ServerService_1.banServerMember)(this.client, this.server.id, this.user.id);
    }
    get roles() {
        return this.roleIds
            .map((id) => this.server.roles.cache.get(id))
            .filter(Boolean);
    }
    permissions() {
        const defaultRoleId = this.server?.defaultRoleId;
        const defaultRole = this.server.roles.cache.get(defaultRoleId);
        let currentPermissions = defaultRole?.permissions || 0;
        const memberRoles = this.roles;
        for (let i = 0; i < memberRoles.length; i++) {
            const role = memberRoles[i];
            currentPermissions = (0, bitwise_1.addBit)(currentPermissions, role?.permissions || 0);
        }
        return currentPermissions;
    }
    hasPermission(permission, ignoreAdmin = false, ignoreCreator = false) {
        if (!ignoreCreator) {
            if (this.server.createdById === this.user.id)
                return true;
        }
        if (!ignoreAdmin) {
            if ((0, bitwise_1.hasBit)(this.permissions(), bitwise_1.RolePermissions.ADMIN))
                return true;
        }
        return (0, bitwise_1.hasBit)(this.permissions(), permission);
    }
}
exports.ServerMember = ServerMember;
class Channels {
    constructor(client) {
        this.client = client;
        this.cache = new Collection();
    }
    setCache(rawChannel) {
        let channel;
        if (rawChannel.serverId)
            channel = new ServerChannel(this.client, rawChannel);
        else
            channel = new Channel(this.client, rawChannel);
        this.cache.set(channel.id, channel);
        return channel;
    }
}
exports.Channels = Channels;
class Channel {
    constructor(client, channel) {
        this.client = client;
        this.id = channel.id;
        this.type = channel.type;
        this.createdAt = channel.createdAt;
        this.lastMessagedAt = channel.lastMessagedAt;
        if (channel.serverId)
            this.server = this.client.servers.cache.get(channel.serverId);
    }
    async send(content, opts) {
        const RawMessage = await (0, MessageService_1.postMessage)({
            client: this.client,
            channelId: this.id,
            content: content,
            silent: opts?.silent,
            nerimityCdnFileId: opts?.nerimityCdnFileId,
            htmlEmbed: opts?.htmlEmbed,
            buttons: opts?.buttons,
            replyToMessageIds: opts?.replyToMessageIds,
            mentionReplies: opts?.mentionReplies,
        });
        const message = new Message(this.client, RawMessage);
        return message;
    }
    toString() {
        return `[#:${this.id}]`;
    }
}
exports.Channel = Channel;
function updateClass(classInstance, update) {
    for (const [key, value] of Object.entries(update)) {
        classInstance[key] = value;
    }
}
class ServerChannel extends Channel {
    constructor(client, channel) {
        super(client, channel);
        this.name = channel.name;
        this.permissions = channel.permissions;
        this.createdById = channel.createdById;
        this.serverId = channel.serverId;
        this.categoryId = channel.categoryId;
        this.server = this.client.servers.cache.get(this.serverId);
    }
}
exports.ServerChannel = ServerChannel;
const UserMentionRegex = /\[@:([0-9]+)\]/gm;
const CommandRegex = /^(\/[^:\s]*):(\d+)( .*)?$/m;
class Message {
    constructor(client, message) {
        this.mentions = [];
        this.client = client;
        this.id = message.id;
        this.channelId = message.channelId;
        this.channel = client.channels.cache.get(this.channelId);
        this.content = message.content;
        this.type = message.type;
        this.createdAt = message.createdAt;
        this.user = this.client.users.cache.get(message.createdBy.id);
        const cmd = message.content?.match(CommandRegex);
        if (cmd?.[2] === this.client.user?.id) {
            this.command = {
                name: cmd?.[1].substring(1),
                args: message.content.split(" ").slice(1),
            };
        }
        if (!this.user) {
            this.user = this.client.users.setCache(message.createdBy);
        }
        if (!this.channel) {
            this.channel = this.client.channels.setCache({
                id: this.channelId,
                type: RawData_1.ChannelType.DM_TEXT,
            });
        }
        if (this.content) {
            const mentionIds = [...this.content.matchAll(UserMentionRegex)].map((exp) => exp[1]);
            this.mentions = mentionIds
                .map((id) => this.client.users.cache.get(id))
                .filter((u) => u);
        }
    }
    get member() {
        return this.channel.server?.members.cache.get(this.user.id);
    }
    reply(content, opts) {
        let fOpts = opts || {};
        fOpts.replyToMessageIds = [this.id];
        return this.channel.send(content, opts);
    }
    async edit(content) {
        const RawMessage = await (0, MessageService_1.editMessage)({
            client: this.client,
            channelId: this.channel.id,
            messageId: this.id,
            content: content,
        });
        const message = new Message(this.client, RawMessage);
        return message;
    }
    async delete() {
        return (0, MessageService_1.deleteMessage)({
            channelId: this.channel.id,
            client: this.client,
            messageId: this.id,
        });
    }
    toString() {
        return `[q:${this.id}]`;
    }
}
exports.Message = Message;
class Posts {
    constructor(client) {
        this.client = client;
    }
    async get(id) {
        const RawPosts = await (0, PostsService_1.getPosts)(this.client);
        const posts = RawPosts.map((post) => new Post(this.client, post));
        return id ? posts.find((p) => p.id === id) : posts;
    }
    async create(content, opts) {
        const RawPost = await (0, PostsService_1.postPost)({
            client: this.client,
            content: content,
            nerimityCdnFileId: opts?.nerimityCdnFileId,
            poll: opts?.poll,
        });
        const post = new Post(this.client, RawPost);
        return post;
    }
}
class Post {
    constructor(client, post) {
        this.client = client;
        this.id = post.id;
        this.content = post.content;
        this.attachments = post.attachments;
        this.deleted = post.deleted;
        this.block = post.block;
        this.commentToId = post.commentToId;
        this.commentTo = post.commentTo;
        this.createdBy = post.createdBy;
        this.createdAt = post.createdAt;
        this.editedAt = post.editedAt;
        this.likedBy = post.likedBy;
        this.reposts = post.reposts;
        this.repost = post.repost;
        this._count = post._count;
        this.views = post.views;
        this.announcement = post.announcement;
        this.poll = post.poll;
    }
    async edit(content) {
        const RawPost = await (0, PostsService_1.editPost)({
            client: this.client,
            content: content,
            postId: this.id,
        });
        const post = new Post(this.client, RawPost);
        return post;
    }
    async delete() {
        await (0, PostsService_1.deletePost)({
            client: this.client,
            postId: this.id,
        });
    }
}
exports.Post = Post;
class User {
    constructor(client, user) {
        this.client = client;
        this.id = user.id;
        this.username = user.username;
        this.tag = user.tag;
        this.hexColor = user.hexColor;
        this.badges = user.badges;
        this.joinedAt = user.joinedAt;
        this.avatar = user.avatar;
        this.banner = user.banner;
        this.bot = user.bot;
    }
    toString() {
        return `[@:${this.id}]`;
    }
}
class ClientUser extends User {
    setActivity(activity) {
        this.client.socket.emit(EventNames_1.SocketClientEvents.UPDATE_ACTIVITY, activity);
    }
    constructor(client, user) {
        super(client, user);
    }
}
class Collection extends Map {
    constructor() {
        super();
    }
}
class Button {
    constructor(client, payload) {
        this.client = client;
        this.id = payload.buttonId;
        this.userId = payload.userId;
        this.channelId = payload.channelId;
        this.messageId = payload.messageId;
        this.user = this.client.users.cache.get(this.userId);
        this.channel = client.channels.cache.get(this.channelId);
        this.data = payload.data;
        this.type = payload.type;
    }
    async respond(opts) {
        await (0, MessageService_1.buttonClickCallback)({
            client: this.client,
            buttonId: this.id,
            channelId: this.channelId,
            messageId: this.messageId,
            userId: this.userId,
            data: opts,
        });
    }
}
exports.Button = Button;
